### DISCUSSION TITLE
Set MAC address at runtime in application

### DISCUSSION THREAD

**Discussion by VineetaNarkhede-eaton** [zephyrproject-rtos/zephyr#39252] [Q&A] [answered]
Hi,
I want to set the MAC address in application, currently it seems that Mac address is being generated in driver file with random generator.
The requirement is that device will have a unique mac address which will be read from external memory at runtime and same will be set forever.
How to set the Mac address in application?
How to set the Mac address independent of device type e.g stm32, esp32 etc. so that same logic would be applicable for any device.

**Comment by jukkar** [comment]
Search `net_if_set_link_addr()` function usage from various Ethernet drivers in `drivers/ethernet`

**Reply by VineetaNarkhede-eaton** [reply]
I found that mac address need to be set at two places 1) dev->Data 2) link_addr
![image](https://user-images.githubusercontent.com/88422239/136530732-cbd0317c-150d-4723-936d-2cc8f968bec7.png)
I am able to use net_if_set_link_addr() function to set link_addr, it also requires setting mac address in dev->Data, but setting mac address in dev->Data nedd access to device information directly as given below, which I want to avoid. And this is becoming very specific to stm32 driver.

1. Get pointer to device data
 const struct device *dev = net_if_get_device(iface)
2. Create a structure to get 4 byte offset to mac address
       struct ctx
	{
		struct net_if *iface; ///< Pointer to network interface
		uint8_t mac_address[6];///< Device MAC address
	};
3. Assign dev->data pointer
       static struct ctx* data = dev->data;
4. copy mac address to data->mac_address
	
So the data is being exposed to end user, How to avoid this direct access to device data which does not seem a good solution to set mac address?

**Comment by jukkar** [comment, marked as answer]
The `dev->data` is the device driver private data, the driver can store the mac address in its context if needed, this is not really relevant to your question. Only way to set the MAC address in the driver so that the rest of the system knows about it is to use `net_if_set_link_addr()`.

I just realized you actually wanted something different, from the application you can do it like this
```
	ret = net_mgmt(NET_REQUEST_ETHERNET_SET_MAC_ADDRESS, iface,
		       &params, sizeof(struct ethernet_req_params));
```
For example usage, see `tests/net/ethernet_mgmt/src/main.c` for details. Note that Ethernet driver needs to implement `set_config` API function and handle the MAC address setting in it, like how it is done for example in `drivers/ethernet/eth_sam_gmac.c`

**Reply by VineetaNarkhede-eaton** [reply]
I tried setting mac address using net_if_set_link_addr() only (not setting addr in dev->data), but it is not working. I am verifying through
net_if_get_link_addr().

```
link_addr = net_if_get_link_addr(iface);
Print link_addr
net_if_set_link_addr(iface, mac_id_string, 6U,  NET_LINK_ETHERNET);
link_addr_updated = net_if_get_link_addr(iface);
Print link_addr_updated 
```

link_addr  and link_addr_updated  are same. 
![image](https://user-images.githubusercontent.com/88422239/136562115-e23a87dc-ff2d-4be2-b120-f07191f66731.png)
I tried with CONFIG_ETH_STM32_HAL_RANDOM_MAC=y and CONFIG_ETH_STM32_HAL_RANDOM_MAC=n, result is same in both cases, link address is not updating. It gets updated only when I update dev->data, could it be a bug or anything else required than above code?

I will look into net_mgmt(....) now.

**Comment by VineetaNarkhede-eaton** [comment]
Hi @jukkar, Thanks for your help. I explored and tried using net_mgmt(..), and found that it updates mac address only if ethernet link is down. 
![image](https://user-images.githubusercontent.com/88422239/137087257-c3f66f80-79e0-4dd5-9629-84a9973de722.png)
I need to update Mac address irrespective of ethernet link status. Could you please suggest how to tackle this problem? is there a way to disable carrier/link even if it's connected and restore later?

**Reply by jukkar** [reply]
The Ethernet link needs to be down when MAC address is changed. Even if Ethernet driver would allow MAC address changing when device is active, the upper network stack needs to clean up the old MAC address use before new address is used.

**Reply by VineetaNarkhede-eaton** [reply]
Thanks @jukkar for clarifying this. Since Ethernet driver is not allowing MAC address changing when link is up, I need to find a way to reset or put the link down even if it's connected. So, I have few questions below,
1. Can we put the link down through any API? 
2. What is the condition required for upper network stack to clean up the old MAC address?

**Reply by jukkar** [reply]
> 1. Can we put the link down through any API?

You can use `int net_if_down(struct net_if *iface);` for that

> 2. What is the condition required for upper network stack to clean up the old MAC address?

Hmm, not sure I understand your question. The most important thing for upper net stack about MAC address, is IPv6 SLAAC where local IPv6 address is constructed from interface MAC address, and ARP handling in IPv4.

**Reply by VineetaNarkhede-eaton** [reply]
Thanks a lot @jukkar
